os/arch/arm/src/imxrt/imxrt_mtd_flash.c

```diff
@@ -0,0 +1,688 @@
+ * 3. Neither the name tinyara nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************************/
+
+/************************************************************************************
+ * Included Files
+ ************************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <sys/types.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/fs/ioctl.h>
+#include <tinyara/spi/spi.h>
+#include <tinyara/fs/mtd.h>
+
+#include "imxrt_flash.h"
+#include "imxrt_log.h"
+
+/************************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************************/
+/* Various manufacturers may have produced the parts.  0x20 is the manufacturer ID
+ * for the STMicro MP25x serial FLASH.  If, for example, you are using the a Macronix
+ * International MX25 serial FLASH, the correct manufacturer ID would be 0xc2.
+ */
+
+#ifndef CONFIG_IMXRT_MANUFACTURER
+//#error  "Manufacturer Id not selected for the flash device"
+#endif
+
+#ifndef CONFIG_IMXRT_MEMORY_TYPE
+//#error  "Memory Type not defined no selected for the flash device"
+#endif
+
+#if defined(CONFIG_ARCH_CHIP_FAMILY_IMXRT102x)
+/*  NorFlash capacity is 8,388,608 bytes:
+ *  (2048 sectors) * (4096 bytes per sector)
+ *  (32768 pages) * (256 bytes per page)
+ */
+#define IMXRT_1020EVK_NORFLASH_SECTOR_SHIFT    12	/* Sector size 1 << 12 = 4096 */
+#define IMXRT_1020EVK_NORFLASH_NSECTORS        2048
+#define IMXRT_1020EVK_NORFLASH_PAGE_SHIFT      8	/* Page size 1 << 8 = 256 */
+#define IMXRT_1020EVK_NORFLASH_NPAGES          32768
+#elif defined(CONFIG_ARCH_CHIP_FAMILY_IMXRT105x)
+/*  HyperFlash capacity is 67,108,864 bytes:
+ *  (100 sectors) * (262,144 bytes per sector)
+ *  (131072 pages) * (512 bytes per page)
+ */
```

Comment:
(Number of sectors) x (bytes per sector) is NOT EQUAL to flash size. 
Can you please explain why the full flash capacity is not used?

Reply:
I am sorry 

I put the value of hexadecimal wrongly.

I fixed it in next update